src/core/qgsunsetattributevalue.h

```diff
@@ -0,0 +1,111 @@
+/***************************************************************************
+                         qgsunsetattributevalue.h
+                         ------------------------
+    begin                : July 2022
+    copyright            : (C) 2022 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSUNSETATTRIBUTEVALUE_H
+#define QGSUNSETATTRIBUTEVALUE_H
+
+#include "qgis_core.h"
+#include <QString>
+#include <QVariant>
+
+/**
+ * \ingroup core
+ * \brief Represents a default, "not-specified" value for a feature attribute.
+ *
+ * \since QGIS 3.28
+ */
+class CORE_EXPORT QgsUnsetAttributeValue
+{
+  public:
+
+    /**
+     * Constructor for a QgsUnsetAttributeValue.
+     */
+    QgsUnsetAttributeValue() = default;
+
+    /**
+     * Constructor for a QgsUnsetAttributeValue, with the specified original provider's default value clause.
+     */
+    explicit QgsUnsetAttributeValue( const QString &defaultValueClause );
+
+    /**
+     * Returns the original data provider's default value clause.
+     */
+    QString defaultValueClause() const { return mDefaultValueClause; }
+
+    /**
+     * QgsUnsetAttributeValues are always considered equal to each other, regardless of what the original provider's default value clause was.
+     */
+    inline bool operator==( const QgsUnsetAttributeValue & ) const { return true; }
```

Comment:
What is a typical situation where this is used/required?
Extrapolating from SQL/NULL logic I'd expect the inverse result.

Reply:
The intention is that QgsUnsetAttributeValue is effectively a static/stateless value. So when we are comparing two features and they both have unset values for a particular attribute, we should always consider these equal (even if they came from two different providers with different default value clauses)).

Reply:
Isn't it very likely that in the end they will not be equal because they will later be populated with different default values?

Reply:
> Isn't it very likely that in the end they will not be equal because they will later be populated with different default values?

Yes -- but in the unsaved state they can still be considered equal. Eg. if I have a new unsaved feature from a gpkg and haven't set the id value, I'd consider that equal to a new unsaved feature from a postgres table where I haven't set the id value (and all the table structure and other field values are the same, of course!).

It's only after you save the feature out and the unset value becomes set that I'd consider them different.

Reply:
> It's only after you save the feature out and the unset value becomes set that I'd consider them different.

I'd rather do the opposite and assume they are not equal until we know better (after saving).
I.e. treat unsetValue like NULL value where SQL goes the pessimistic approach assuming `NULL != NULL`.

Reply:
Hmm... I'm worried that would break a lot of things

e.g. check this precedent:

![image](https://user-images.githubusercontent.com/1829991/179685486-5e2f8fac-17cb-4f7a-af77-338541bf9969.png)

:wink: 

I realise this is because NULL is a "singleton" in PyQGIS land. I'd like this to be treated similar.

Reply:
The precedent here was created by me, reasoning follows :wink: 

I originally wanted to have it report `NULL is NULL` as `true` and `NULL == NULL` as `false` but often API would return `None` for `feat.attribute(3)` and it was technically not possible to have `None is NULL` return `True`, so the alternative was to resort to `==` where we do have the control over return value.
I was hoping that now would be a chance to fix that...

Reply:
Just to mention another reference besides SQL (before going quiet) is [the C++ ref for NaN](https://stackoverflow.com/a/15268951/2319028) which also handles this unknown value pessimistically as not equal.
But if this leads to hard-to-fix-problems with the current codebase, then this can be ignored.

Reply:
Given that we're mostly(?) focused on the Python aspect here...

Python doesn't treat None like a SQL NULL value. E.g:

    >>>    [1,2,None,3] == [1,2,None,3]
    True

I really feel like we'll bump up against a lot of unexpected behaviour if we force `Unset != Unset`. I.e:

    >>> [1,2,QgsUnsetAttributeValue(), 3] == [1,2,QgsUnsetAttributeValue(), 3]
    False

is very unpredictable given the equality of the lists with None values instead. For example, let's say we have some code like this:

```
feature = next(layer.getFeatures())
# pretend at this stage that the attributes for feature are [1,2,QgsUnsetAttributeValue(), 3]

cleaned_feature = clean_up_feature(feature)
# where clean_up_feature is a function which may or may not modify a feature

assert cleaned_feature.attributes() != feature.attributes(), 'Feature attributes were not clean!!!'
```

Because of the inequality of the unset value attribute, the assert will always be hit in this circumstance, even if the function doesn't do anything to the feature. And it's going to be a nightmare for a (mere mortal) developer to debug :scream: 

I guess taking a few steps back I'm just not convinced that mimic SQL behaviour makes sense in a non-SQL environment.

Reply:
Valid points, coming from to-be-expected real-world problems.

Where my concern is coming from is that features that are at a certain point in time considered equal suddenly may later be non-equal.

E.g. consider this:

```py
feature1 = QgsFeature( fid: undefined, amount: 3 )
feature2 = QgsFeature( fid: undefined, amount: 3 )
feature1 == feature2; # True
layer.dataProvider().addFeature(feature1)
layer.dataProvider().addFeature(feature2)
feature1 == feature2 # False 😱
```

Reply:
> layer.dataProvider().addFeature(feature1)
> layer.dataProvider().addFeature(feature2)

Hmm -- but even without an unset value this is likely to be the case, because the data provider will set different feature IDs on these two features. (unless you set the FastInsert flag, which would also mean that the unset value attribute won't get updated either)

So a better example would be:

```
feature1.attributes() == feature2.attributes(); # True
layer.dataProvider().addFeature(feature1)
layer.dataProvider().addFeature(feature2)
feature1.attributes() == feature2.attributes() # False 😱
```

but I'm personally ok with this situation, because the addFeature method doesn't advertise that it won't modify the feature in place. And it's easy enough for a developer to debug this one, just by visually comparing the contents of the feature1.attributes() vs feature2.attributes() after the data provider calls they'll clearly see that the values are different.

Reply:
Convinced. It's easier to debug, will likely avoid confused mails and issues
:+1:

Reply:
Thanks Matthias -- I really appreciate being able to brainstorm this with you!